synapse/rest/client/v2_alpha/register.py

```diff
@@ -377,6 +377,33 @@ async def on_GET(self, request):
             return 200, {"available": True}
 
 
+class RegistrationTokenValidityRestServlet(RestServlet):
```

Comment:
I'm a bit confused by all the rate limiting stuff, do I need to add that here somewhere?

Reply:
Yep, you'll want to do so (otherwise people may be tempted to try and brute force a token).

The `Ratelimiter` class is indeed a bit complicated, but calling it should be fairly straight-forward. Since we don't have a user account to rate-limit on yet, we'll probably need to rate-limit by IP address. Thus, we just need to grab the IP address of the requester and use that as the ratelimit key:

```python
        # Somewhere in __init__()
        self._my_ratelimiter = Ratelimiter(
            store=hs.get_datastore(),
            clock=hs.get_clock(),
            rate_hz=hs.config.ratelimiting.rc_your_ratelimiter.per_second,
            burst_count=hs.config.ratelimiting.rc_your_ratelimiter.burst_count,
        )

        # In your on_GET
        await self._my_ratelimier.ratelimit(
            None, (request.getClientIP(),)
        )
```

`ratelimit` will raise a `LimitExceededError` if the limit has been reached, and will inform the client on the `ms` they need to wait before they can try again.

The other question is whether to create a new ratelimiting config option for this endpoint, or to just reuse `rc_registration`. Personally I think it's always best to separate config options out in case people want to tweak them individually (plus, it would be annoying to get your token wrong a few times before eventually coming to hit `/register` and being denied because of the previous token checks).

Ratelimiting configs are managed by the [RatelimitConfig](https://github.com/matrix-org/synapse/blob/2ca4e349e9d0c606d802ae15c06089080fa4f27e/synapse/config/ratelimiting.py#L46) class.

Reply:
Thanks!

Reply:
Ok, I've added that. Not sure if I got the defaults right.
https://github.com/matrix-org/synapse/pull/10142/commits/003e67d1ab5d64ea54f3c125ccd2b7a03e18cddc

Reply:
I think the current defaults are fine, and the commit looks good otherwise as well!

It's good to leave a little bit of a burst count on IP rate-limiting, in case you have a situation where a school tries to sign up a whole class at once or something.